internal/usecase/repository/hostconf_jwk_repository.go

```diff
@@ -17,77 +17,140 @@ var notImplementedError = fmt.Errorf("TODO: not implemented")
 
 type hostconfJwkRepository struct {
 	config *config.Config
-	// TODO: store JWKs in database
-	signingKeys []jwk.Key
-	publicKeys  []string
 }
 
 func NewHostconfJwkRepository(cfg *config.Config) repository.HostconfJwkRepository {
 	if cfg == nil {
 		panic("'cfg' is nil")
 	}
 	r := &hostconfJwkRepository{config: cfg}
-	err := r.generateEphemeralJWK()
-	if err != nil {
-		panic(err)
-	}
 	return r
 }
 
 // CreateJWK generates and inserts a new JWK into the database. The private
 // key is encrypted with the current app secret.
-func (r *hostconfJwkRepository) CreateJWK(db *gorm.DB) (model *model.HostconfJwk, err error) {
-	return nil, notImplementedError
+func (r *hostconfJwkRepository) InsertJWK(db *gorm.DB, hcjwk *model.HostconfJwk) (err error) {
+	if db == nil {
+		return internal_errors.NilArgError("db")
+	}
+
+	if err = db.Create(&hcjwk).Error; err != nil {
+		return err
+	}
+	return nil
 }
 
 // RevokeJWK revokes a JWK with key identifier `kid`
-func (r *hostconfJwkRepository) RevokeJWK(db *gorm.DB, kid string) (model *model.HostconfJwk, err error) {
-	return nil, notImplementedError
+func (r *hostconfJwkRepository) RevokeJWK(db *gorm.DB, kid string) (hcjwk *model.HostconfJwk, err error) {
+	if db == nil {
+		return nil, internal_errors.NilArgError("db")
+	}
+	// find JKW by unique kid
+	if err = db.
+		Where("key_id = ?", kid).
+		First(&hcjwk).
+		Error; err != nil {
+		return nil, err
+	}
+
+	// Set encrypted JWK field to NULL and safe the new object
+	hcjwk.EncryptedJwk = nil
+	if err = db.Save(hcjwk).Error; err != nil {
+		return nil, err
+	}
+
+	return hcjwk, nil
 }
 
 // ListJWKs all JWKs, including expired and revoked JWKs
-func (r *hostconfJwkRepository) ListJWKs(db *gorm.DB) (models []model.HostconfJwk, err error) {
-	return nil, notImplementedError
+func (r *hostconfJwkRepository) ListJWKs(db *gorm.DB) (hcjwks []model.HostconfJwk, err error) {
+	if db == nil {
+		return nil, internal_errors.NilArgError("db")
+	}
+	// list JWKs, order by id to have determinstic sorting
+	if err = db.
+		Order("id").
+		Find(&hcjwks).
+		Error; err != nil {
+		return nil, err
+	}
+	return hcjwks, nil
+}
+
+// PurgeExpiredJWKs find and removes all JWKs that are expired
+func (r *hostconfJwkRepository) PurgeExpiredJWKs(db *gorm.DB) (hcjwks []model.HostconfJwk, err error) {
+	if db == nil {
+		return nil, internal_errors.NilArgError("db")
+	}
+
+	// list JWKs, order by id to have determinstic sorting
+	now := time.Now()
+	if err = db.
+		Order("id").
+		Where("expires_at <= ?", now). // use SQL NOW()?
```

Comment:
Should we allow for a small clock skew here?  We should be generating and switching to a new key ~30 days before expiry, but just in case?

Reply:
A small clock skew would make testing more complicated. Besides an hour clock skew doesn't make a difference when the renewal window is about a month.

Reply:
OK, I'm satisfied.